src/natpmp/gateway.cpp

```diff
@@ -0,0 +1,548 @@
+#define USE_SOCKET_ROUTE
+#undef USE_SYSCTL_NET_ROUTE
+#endif
+
+#ifdef WIN32
+#undef USE_PROC_NET_ROUTE
+#undef USE_SOCKET_ROUTE
+#undef USE_SYSCTL_NET_ROUTE
+//#define USE_WIN32_CODE
+#define USE_WIN32_CODE_2
+#endif
+
+#ifdef __CYGWIN__
+#undef USE_PROC_NET_ROUTE
+#undef USE_SOCKET_ROUTE
+#undef USE_SYSCTL_NET_ROUTE
+#define USE_WIN32_CODE
+#include <stdarg.h>
+#include <w32api/windef.h>
+#include <w32api/winbase.h>
+#include <w32api/winreg.h>
+#endif
+
+#ifdef __HAIKU__
+#include <sys/sockio.h>
+#define USE_HAIKU_CODE
+#endif
+
+#ifdef USE_SYSCTL_NET_ROUTE
+#include <sys/sysctl.h>
+#endif
+
+#ifdef USE_SOCKET_ROUTE
+#include <string.h>
+#endif
+
+#if defined(USE_SOCKET_ROUTE) || defined(__HAIKU__)
+#include <unistd.h>
+#include <net/if.h>
+#endif
+
+#if defined(USE_SYSCTL_NET_ROUTE) || defined(__HAIKU__)
+#include <stdlib.h>
+#endif
+
+#if defined(USE_SOCKET_ROUTE) || defined(USE_SYSCTL_NET_ROUTE)
+#include <sys/socket.h>
+#include <net/route.h>
+#endif
+
+#ifdef USE_WIN32_CODE
+#include <unknwn.h>
+#include <winreg.h>
+#define MAX_KEY_LENGTH 255
+#define MAX_VALUE_LENGTH 16383
+#endif
+
+#ifdef USE_WIN32_CODE_2
+#include <windows.h>
+#include <iphlpapi.h>
+#endif
+#include <gateway.h>
+#ifndef WIN32
+#define SUCCESS (0)
+#define FAILED  (-1)
+#endif
+#ifdef USE_PROC_NET_ROUTE
+#include <sstream>
+#include <list>
+#include <algorithm>
+/*
+ * There is no portable method to get the default route gateway.
+ * So below are four (or five ?) different functions implementing this.
+ * Parsing /proc/net/route is for Linux.
+ * sysctl is the way to access such information on BSD systems.
+ * Many systems should provide route information through raw PF_ROUTE
+ * sockets.
+ * In MS Windows, default gateway is found by looking into the registry
+ * or by using GetBestRoute().
+ *
+ */
+
+/*
+ parse /proc/net/route which is as follow :
+
+Iface   Destination     Gateway         Flags   RefCnt  Use     Metric  Mask            MTU     Window  IRTT
+wlan0   0001A8C0        00000000        0001    0       0       0       00FFFFFF        0       0       0
+eth0    0000FEA9        00000000        0001    0       0       0       0000FFFF        0       0       0
+wlan0   00000000        0101A8C0        0003    0       0       0       00000000        0       0       0
+eth0    00000000        00000000        0001    0       0       1000    00000000        0       0       0
+
+ One header line, and then one line by route by route table entry.
+*/
+
+int TestIfWhiteSpace(char character)
+{
+    std::list<char> whiteSpaces = {'\f', '\n', '\r', '\t', '\v'};
+    return std::count(whiteSpaces.begin(), whiteSpaces.end(), character);
+}
+
+int GetDefaultGateway(in_addr_t * addr)
+{
+        int64_t d = 0;
+        int64_t g = 0;
+        char buf[256] = {0};
+        int line = 0;
+        char * p = nullptr;
+        FILE* f = fopen("/proc/net/route", "r");
+        if(!f)
+                return FAILED;
+        while(fgets(buf, sizeof(buf), f)) {
```

Comment:
Consider running `clang-tidy` on all new files. For changes to existing files: use https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/clang-format-diff.py